### PR TITLE
Add SSL/TLS support

### PR DESCRIPTION
**Pull request by valashko**
Tested both SSL and TLS without certificate verification, i.e. `verifyMode = CVerifyNone`. Details are described in *tests/test_wss.nim*

### CODE REVIEW

File: src/news.nim [hunk 1]
```diff
@@ -114,43 +114,65 @@ when not newsUseChronos:
...
     return ws
 
-proc newWebSocket*(url: string, headers: StringTableRef = nil): Future[WebSocket] {.async.} =
+proc newWebSocket*(url: string, headers: StringTableRef = nil,
+                   protVersion: SslProtVersion = protSSLv23,
+                   verifyMode: SslCVerifyMode = CVerifyPeer): Future[WebSocket] {.async.} =
   ## Creates a client
   var ws = WebSocket()
-  let uri = parseUri(url)
+  var uri = parseUri(url)
```
Comment (yglukhov) [inline src/news.nim]: I can't see where the `uri.scheme` change is required, thus no need for this?

File: src/news.nim [hunk 2]
```diff
@@ -114,43 +114,65 @@ when not newsUseChronos:
...
-    raise newException(WebSocketError, &"Scheme {uri.scheme} not supported yet.")
+  var useSsl = false
+  case uri.scheme
+    of "wss":
+      uri.scheme = "https"
+      port = Port(443)
+      useSsl = true
+    of "ws":
+      uri.scheme = "http"
+      port = Port(80)
```
Comment (yglukhov) [inline src/news.nim]: `port` is already 80 at this point. No need reassigning.

File: src/news.nim [hunk 3]
```diff
@@ -114,43 +114,65 @@ when not newsUseChronos:
...
   var ws = WebSocket()
-  let uri = parseUri(url)
+  var uri = parseUri(url)
   var port = Port(80)
-  if uri.scheme != "ws":
-    raise newException(WebSocketError, &"Scheme {uri.scheme} not supported yet.")
+  var useSsl = false
+  case uri.scheme
+    of "wss":
+      uri.scheme = "https"
```
Comment (yglukhov) [inline src/news.nim]: No need?

File: src/news.nim [hunk 4]
```diff
@@ -114,43 +114,65 @@ when not newsUseChronos:
...
+      port = Port(443)
+      useSsl = true
+    of "ws":
+      uri.scheme = "http"
```
Comment (yglukhov) [inline src/news.nim]: No need?

File: src/news.nim [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import strutils, streams, random, base64, uri, strformat, nativesockets, oids,
-  strtabs
+  strtabs, std/sha1
```
Comment (yglukhov) [inline src/news.nim]: Why is `sha1` needed?

File: src/news.nim [hunk 6]
```diff
@@ -114,43 +114,65 @@ when not newsUseChronos:
...
+      raise newException(WebSocketError, &"Scheme {uri.scheme} not supported yet.")
   if uri.port.len > 0:
     port = Port(parseInt(uri.port))
 
   when newsUseChronos:
     ws.transp = await connect(resolveTAddress(uri.hostname, port)[0])
   else:
     ws.transp = newAsyncSocket()
+    if useSsl:
+      let sslCtx = newContext(protVersion = protVersion, verifyMode = verifyMode)
```
Comment (yglukhov) [inline src/news.nim]: The context should be deleted when done. First solution which comes to mind is to keep it as a field.

File: src/news.nim [hunk 7]
```diff
@@ -114,43 +114,65 @@ when not newsUseChronos:
...
-      hello &= "\c\L"
-  hello &= "\c\L"
+      hello &= CRLF
+  hello &= CRLF
 
   await ws.transp.send(hello)
 
   var output = ""
-  while not output.endsWith("\c\L\c\L"):
+  while not output.endsWith(CRLF & CRLF):
```
Comment (yglukhov) [inline src/news.nim]: `CRLF & CRLF` would be evaluating and allocating on every cycle. Use `static(CRLF & CRLF)` to force evaluation to compile time, or introduce a `let delimiter = CRLF & CRLF` or smth.

File: src/news.nim (hunk 5 above)
Comment (valashko) [reply src/news.nim]: The code does not compile without it `news\src\news.nim(100, 14) Error: undeclared identifier: 'secureHash'`

File: src/news.nim (hunk 1 above)
Comment (valashko) [reply src/news.nim]: The change was ported from the orignal repository ([see this commit](https://github.com/treeform/ws/commit/5c5f6ae0e2afd9078f1eb093e273948205dc37ca)). You are right. The code works fine without it.

File: src/news.nim (hunk 2 above)
Comment (valashko) [reply src/news.nim]: Sure

File: src/news.nim (hunk 3 above)
Comment (valashko) [reply src/news.nim]: The change was ported from the orignal repository ([see this commit](https://github.com/treeform/ws/commit/5c5f6ae0e2afd9078f1eb093e273948205dc37ca)). You are right. The code works fine without it.

File: src/news.nim (hunk 4 above)
Comment (valashko) [reply src/news.nim]: The change was ported from the orignal repository ([see this commit](https://github.com/treeform/ws/commit/5c5f6ae0e2afd9078f1eb093e273948205dc37ca)). You are right. The code works fine without it.

File: src/news.nim (hunk 7 above)
Comment (valashko) [reply src/news.nim]: ok

File: src/news.nim (hunk 6 above)
Comment (valashko) [reply src/news.nim]: fixed by c6137b9

File: src/news.nim (hunk 4 above)
Comment (valashko) [reply src/news.nim]: fixed by c6137b9

File: src/news.nim (hunk 3 above)
Comment (valashko) [reply src/news.nim]: fixed by c6137b9

File: src/news.nim (hunk 2 above)
Comment (valashko) [reply src/news.nim]: fixed by c6137b9

File: src/news.nim (hunk 1 above)
Comment (valashko) [reply src/news.nim]: fixed by c6137b9

File: src/news.nim (hunk 7 above)
Comment (valashko) [reply src/news.nim]: fixed by c6137b9

File: src/news.nim [hunk 8]
```diff
@@ -114,43 +126,60 @@ when not newsUseChronos:
...
     for k, v in headers:
-      hello &= k
-      hello &= ": "
-      hello &= v
-      hello &= "\c\L"
-  hello &= "\c\L"
+      customHeaders &= k
+      customHeaders &= ": "
+      customHeaders &= v
+      customHeaders &= CRLF
```
Comment (Tormund) [inline src/news.nim]: customHeaders &= &"{k}:{v}{CRLF}"

Comment (Tormund) [reply src/news.nim]: customHeaders &= "$#: $#$#" % [k, v, CRLF]

Comment (valashko) [reply src/news.nim]: fixed

**Pull request merged**